src/main/java/frc/utils/interpolator/InterpolationUtils.java

```diff
@@ -0,0 +1,149 @@
+			xMin = Math.min(xMin, point.getFirst().getX());
+			xMax = Math.max(xMax, point.getFirst().getX());
+			yMin = Math.min(yMin, point.getFirst().getY());
+			yMax = Math.max(yMax, point.getFirst().getY());
+		}
+		return query.getX() >= xMin && query.getX() <= xMax && query.getY() >= yMin && query.getY() <= yMax;
+	}
+
+	@SafeVarargs
+	public static Pair<Translation2d, Double>[] createBoundingBox(Translation2d query, Pair<Translation2d, Double>... knownPoints) {
+		Pair<Translation2d, Double> bottomLeft = null, bottomRight = null, topLeft = null, topRight = null;
+
+		for (Pair<Translation2d, Double> point : knownPoints) {
+			double x = point.getFirst().getX();
+			double y = point.getFirst().getY();
+			if (x <= query.getX() && y <= query.getY()) {
+				if (
+					bottomLeft == null
+						|| (x >= bottomLeft.getFirst().getX() || y >= bottomLeft.getFirst().getY())
+							&& Translation2dUtils
+								.isClosestPointFromPoints(query, point.getFirst(), new Translation2d[] {bottomLeft.getFirst(), point.getFirst()})
+				) {
+					bottomLeft = point;
+				}
+			}
+			if (x >= query.getX() && y <= query.getY()) {
+				if (
+					bottomRight == null
+						|| x >= bottomRight.getFirst().getX()
+						|| y <= bottomRight.getFirst().getY()
+							&& Translation2dUtils.isClosestPointFromPoints(
+								query,
+								point.getFirst(),
+								new Translation2d[] {bottomRight.getFirst(), point.getFirst()}
+							)
+				) {
+					bottomRight = point;
+				}
+			}
+			if (x <= query.getX() && y >= query.getY()) {
+				if (
+					topLeft == null
+						|| x >= topLeft.getFirst().getX()
+						|| y <= topLeft.getFirst().getY()
+							&& Translation2dUtils
+								.isClosestPointFromPoints(query, point.getFirst(), new Translation2d[] {topLeft.getFirst(), point.getFirst()})
+				) {
+					topLeft = point;
+				}
+			}
+			if (x >= query.getX() && y >= query.getY()) {
+				if (
+					topRight == null
+						|| x <= topRight.getFirst().getX()
+						|| y <= topRight.getFirst().getY()
+							&& Translation2dUtils
+								.isClosestPointFromPoints(query, point.getFirst(), new Translation2d[] {topRight.getFirst(), point.getFirst()})
+				) {
+					topRight = point;
+				}
+			}
+		}
+		return new Pair[] {bottomLeft, bottomRight, topLeft, topRight};
+	}
+
+	public static double biLinearInterpolate(
+		Pair<Translation2d, Double> bottomLeft,
+		Pair<Translation2d, Double> bottomRight,
+		Pair<Translation2d, Double> topLeft,
+		Pair<Translation2d, Double> topRight,
+		Translation2d query
+	) {
+		double x1 = bottomLeft.getFirst().getX(), y1 = bottomLeft.getFirst().getY();
+		double x2 = bottomRight.getFirst().getX(), y2 = topLeft.getFirst().getY();
+
+		double x = query.getX();
+		double y = query.getY();
+
+		return (bottomLeft.getSecond() * (x2 - x) * (y2 - y)
+			+ bottomRight.getSecond() * (x - x1) * (y2 - y)
+			+ topLeft.getSecond() * (x2 - x) * (y - y1)
+			+ topRight.getSecond() * (x - x1) * (y - y1)) / ((x2 - x1) * (y2 - y1));
```

Comment:
what is this??
maybe save some differences or rename these variables

Reply:
its a formula
![P6TFm](https://github.com/user-attachments/assets/829122d4-6b20-4a77-9351-954e199edefa)

Reply:
so i took out the (x2-x1)/(y2-y1) out
like (a+b)(a) = a^2 + ab

Reply:
ok but also make it like this

return (bottomLeft.getSecond() * (x2 - x) * (y2 - y)
			+ bottomRight.getSecond() * (x - x1) * (y2 - y)
			+ topLeft.getSecond() * (x2 - x) * (y - y1)
			+ topRight.getSecond() * (x - x1) * (y - y1))
           / ((x2 - x1) * (y2 - y1));